src/SmartRange.ts

```
@@ -0,0 +1,207 @@
+        //  TODO accept options for excludeStart, includeEnd
+        if (
+            !Number.isInteger(start) || !Number.isInteger(end) || step
+                ? !Number.isInteger(step)
+                : false
+        ) {
+            throw new TypeError(
+                "SmartRange: start, end and step must be integers"
+            );
+        }
+        this.#start = start;
+        this.#end = end;
+        this.#step =
+            step === undefined || step === 0 ? this.#defaultStep : step;
+    }
+
+    get start(): number {
+        return this.#start;
+    }
+    get end(): number {
+        return this.#end;
+    }
+    get step(): number {
+        return this.#step;
+    }
+    get length(): number {
+        if (this.#isInvalid) return 0;
+        const res = Math.abs(this.#end - this.#start) / this.#step;
+        return res < 0 ? Math.floor(res) : Math.ceil(res);
+    }
+    get #defaultStep(): number {
+        return this.#start < this.#end ? 1 : -1;
+    }
+    get #isInvalid(): boolean {
+        const cond1 = this.#start < this.#end && this.#step < 0;
+        const cond2 = this.#end < this.#start && 0 < this.#step;
+        return cond1 || cond2;
+    }
+
+    set start(v: number) {
+        if (this.#start === v) return;
+        if (!Number.isInteger(v))
+            throw new TypeError("SmartRange: start needs to be integer");
+        this.#start = v;
+    }
+    set end(v: number) {
+        if (this.#end === v) return;
+        if (!Number.isInteger(v))
+            throw new TypeError("SmartRange: end needs to be integer");
+        this.#end = v;
+    }
+    set step(v: number | undefined) {
+        if (this.step === v) return;
+        if (!Number.isInteger(v))
+            throw new TypeError("SmartRange: step needs to be integer");
+        // FIXME what happens to this.#doneSteps when step size is changed after steps are already done
+        if (this.#doneSteps)
+            console.warn(
```

Printing a warning for undefined behaviour is not the expected behaviour imo

---

There is no expected behaviour at this point, since it is not tested, hence the warning 😆 Ideally I'll figure something out for the 2nd release.

---

That's okay but I think if you know about a condition that could be potencially dangerous (even it looks like this scenario can never occurr) you should at least throw an exception. Warnings can be ignored and then nobody recognizes the errorious behaviour.

---

I see that point. I'll change it for now to throw 👍🏻